File: resources/recipes/data/parse-toml-files/parse_toml.js
// Parse and write TOML files in JavaScript.
// Requires: npm install @iarna/toml
import toml from "@iarna/toml";
import fs from "fs";

function readToml(path) {
  const content = fs.readFileSync(path, "utf8");
  return toml.parse(content);
}

function writeToml(data, path) {
  const content = toml.stringify(data);
  fs.writeFileSync(path, content, "utf8");
}

function deepMerge(base, override) {
  const result = { ...base };
  for (const [key, value] of Object.entries(override)) {
    if (
      key in result &&
      typeof result[key] === "object" &&
      !Array.isArray(result[key]) &&
      typeof value === "object" &&
      !Array.isArray(value)
    ) {
      result[key] = deepMerge(result[key], value);
    } else {
      result[key] = value;
    }
  }
  return result;
}

// Example usage
const config = readToml("config.toml");
console.log(config.app.name);
console.log(config.database.host);
console.log(config.servers?.length ?? 0);

File: resources/recipes/data/parse-toml-files/parse_toml.py
"""Parse and write TOML files in Python.

Requires Python 3.11+ (tomllib in stdlib) or tomli for older versions.
For writing: pip install tomli-w
"""
import tomllib
from pathlib import Path


def read_toml(path: str) -> dict:
    """Read a TOML file and return a dictionary."""
    with open(path, "rb") as f:
        return tomllib.load(f)


def write_toml(data: dict, path: str) -> None:
    """Write a dictionary to a TOML file. Requires tomli-w."""
    import tomli_w
    with open(path, "wb") as f:
        tomli_w.dump(data, f)


def merge_configs(base_path: str, env: str = "dev") -> dict:
    """Load base config and merge environment-specific overrides."""
    base = tomllib.loads(Path(base_path).read_text())
    env_file = Path(f"config/{env}.toml")
    if env_file.exists():
        override = tomllib.loads(env_file.read_text())
        return _deep_merge(base, override)
    return base


def _deep_merge(base: dict, override: dict) -> dict:
    result = base.copy()
    for key, value in override.items():
        if key in result and isinstance(result[key], dict) and isinstance(value, dict):
            result[key] = _deep_merge(result[key], value)
        else:
            result[key] = value
    return result


if __name__ == "__main__":
    config = read_toml("config.toml")
    print(config["app"]["name"])
    print(config["database"]["host"])
    print(len(config.get("servers", [])))
